zipkin-lens/src/components/DiscoverPage/SearchBar/SearchBar.tsx

```diff
@@ -0,0 +1,262 @@
+/*
+ * Copyright 2015-2020 The OpenZipkin Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+import { faPlus } from '@fortawesome/free-solid-svg-icons';
+import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
+import {
+  Box,
+  Button,
+  Theme,
+  createStyles,
+  makeStyles,
+} from '@material-ui/core';
+import React, { useCallback, useEffect, useRef, useState } from 'react';
+import { connect } from 'react-redux';
+import { ThunkDispatch } from 'redux-thunk';
+
+import Criterion, { newCriterion } from '../Criterion';
+import CriterionBox from './CriterionBox';
+import { fetchAutocompleteValues } from '../../../actions/autocomplete-values-action';
+import { fetchRemoteServices } from '../../../actions/remote-services-action';
+import { fetchServices } from '../../../actions/services-action';
+import { fetchSpans } from '../../../actions/spans-action';
+import RootState from '../../../types/RootState';
+
+const useStyles = makeStyles((theme: Theme) =>
+  createStyles({
+    addButton: {
+      height: 40,
+      width: 40,
+      minWidth: 40,
+      color: theme.palette.common.white,
+    },
+  }),
+);
+
+type SearchBarProps = {
+  searchTraces: () => void;
+  criteria: Criterion[];
+  onChange: (criteria: Criterion[]) => void;
+  serviceNames: string[];
+  isLoadingServiceNames: boolean;
+  spanNames: string[];
+  isLoadingSpanNames: boolean;
+  remoteServiceNames: string[];
+  isLoadingRemoteServiceNames: boolean;
+  autocompleteKeys: string[];
+  autocompleteValues: string[];
+  isLoadingAutocompleteValues: boolean;
+  loadServices: () => void;
+  loadRemoteServices: (serviceName: string) => void;
+  loadSpans: (serviceName: string) => void;
+  loadAutocompleteValues: (autocompleteKey: string) => void;
+};
+
+export const SearchBarImpl: React.FC<SearchBarProps> = ({
+  searchTraces,
+  criteria,
+  onChange,
+  serviceNames,
+  isLoadingServiceNames,
+  spanNames,
+  isLoadingSpanNames,
+  remoteServiceNames,
+  isLoadingRemoteServiceNames,
+  autocompleteKeys,
+  autocompleteValues,
+  isLoadingAutocompleteValues,
+  loadServices,
+  loadRemoteServices,
+  loadSpans,
+  loadAutocompleteValues,
+}) => {
+  const classes = useStyles();
+
+  // criterionIndex is the index of the criterion currently being edited.
+  // If the value is -1, there is no criterion being edited.
+  const [criterionIndex, setCriterionIndex] = useState(-1);
+
+  const handleCriterionFocus = (index: number) => () => {
```

Comment:
As far as I can tell this callback pattern has bad performance since the functions all get initialized every render. Instead of using the `index` factory can't you do something like `useCallback((criterion) => criteria.find(c => c.id == criterion.id)`?